src/modules/adaptors/Balancer/BalancerPoolAdaptor.sol

```diff
@@ -0,0 +1,295 @@
+     * @return address adhereing to IVault
+     */
+    function vault() internal pure returns (IVault) {
+        return IVault(0xBA12222222228d8Ba445958a75a0704d566BF2C8);
+    }
+
+    // /**
+    //  * @notice the BalancerQueries address on all networks
+    //  * @return address adhering to IBalancerQueries
+    //  */
+    // function balancerQueries() internal pure returns (IBalancerQueries) {
+    //     return 0xE39B5e3B6D74016b2F6A9673D7d7493B6DF549d5;
+    // }
+
+    function relayer() internal pure returns (IBalancerRelayer relayer) {
+        return IBalancerRelayer(0xfeA793Aa415061C483D2390414275AD314B3F621);
+    }
+
+    function gaugeFactory() internal pure returns (ILiquidityGaugeFactory gaugeFactory) {
+        return ILiquidityGaugeFactory(0x4E7bBd911cf1EFa442BC1b2e9Ea01ffE785412EC);
+    }
+
+    //==================== TODO: Adaptor Data Specification ====================
+    // See Related Open Issues on this for BalancerPoolAdaptor.sol
+    //================= Configuration Data Specification =================
+    // NOT USED
+    //====================================================================
+
+    /**
+     @notice bptOut lower than desired
+     */
+    error BalancerPoolAdaptor__BPTOutTooLow();
+
+    /**
+     @notice no claimable reward tokens
+     */
+    error BalancerPoolAdaptor__ZeroClaimableRewards();
+
+    /**
+     * @notice max slippage exceeded
+     */
+    error BalancerPoolAdaptor__MaxSlippageExceeded();
+
+    /**
+     * @notice bpt amount to unstake/withdraw requested exceeds amount actuall staked/deposited in liquidity gauge
+     */
+    error BalancerPoolAdaptor__NotEnoughToWithdraw();
+
+    /**
+     * @notice bpt amount to unstake/withdraw requested exceeds amount actuall staked/deposited in liquidity gauge
+     */
+    error BalancerPoolAdaptor__NotEnoughToDeposit();
+
+    //============================================ Global Functions ===========================================
+    /**
+     * @dev Identifier unique to this adaptor for a shared registry.
+     * Normally the identifier would just be the address of this contract, but this identifier is needed during Cellar Delegate Call Operations, so getting the address of the adaptor is more difficult.
+     * @return encoded adaptor identifier
+     */
+    function identifier() public pure override returns (bytes32) {
+        return keccak256(abi.encode("Balancer Pool Adaptor V 1.0"));
+    }
+
+    //============================================ Implement Base Functions ===========================================
+
+    /**
+     * @notice User deposits are NOT allowed into this position.
+     */
+    function deposit(uint256, bytes memory, bytes memory) public pure override {
+        revert BaseAdaptor__UserDepositsNotAllowed();
+    }
+
+    /**
+     * @notice User withdraws are NOT allowed from this position.
+     */
+    function withdraw(uint256, address, bytes memory, bytes memory) public pure override {
```

Comment:
If you want to try to tackle this feel free, otherwise we can do it together when I am back, but ideally we can allow user withdraws from balancer positions.

To support this, we'll need to implement `withdraw` and `withdrawableFrom`

`withdrawableFrom` will be basically returning the same thing as `balanceOf`.

`withdraw` will need to run a `blockExternalRecevier` check(see other adaptors, like Aave A Token), then if the amount is greater than the amount of BPT the cellar is currently holding in its wallet, it will need to withdraw some BPTs from the guage, then transfer the bpts to the user address provided. Also probs want to have the withdraw NOT harvest rewards since it will add extra gas to the users withdraw TX, and its not really benefitting the strategist much cuz they will probs be calling claim rewards again when rebalancing

Reply:
Kk I'll look into this as well and update this PR with a message once it is done. The thing I'm wondering about though is whether or not we want users to be able to withdraw directly from this `Adaptor` contract. They should be able to call `redeem()` or `withdraw()` from the `Cellar` right? That will do the functionality that you mention I thought.